src/storage/page/page_guard.cpp

```diff
@@ -3,6 +3,30 @@
 
 namespace bustub {
 
+auto BasicPageGuard::UpgradeRead() -> ReadPageGuard {
+  // Save the current page id & bpm (This is because calling Drop() will set these to invalid)
+  auto *bpm = bpm_;
+  auto cur_pid = page_->GetPageId();
+  // First drop the current page
+  this->Drop();
+  // Fetch a new ReadPageGuard from the previously saved BPM
+  auto ret_pg = bpm->FetchPageRead(cur_pid);
+  // Return the rvalue, will invoke move assignment operator
+  return ret_pg;
+}
+
+auto BasicPageGuard::UpgradeWrite() -> WritePageGuard {
+  // Save the current page id & bpm (Same as read)
+  auto *bpm = bpm_;
+  auto cur_pid = page_->GetPageId();
+  // First drop the current page guard
+  this->Drop();
```

Comment:
In this implementation, there is possibility that the page will be evicted between L23 and L25?

Reply:
Ahh I see, I ignore the inner condition when implementing the upgrade logic😭